presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -72,7 +72,11 @@
     private final ScheduledExecutorService yieldExecutor;
     private final long maxSpill;
     private final SpillSpaceTracker spillSpaceTracker;
-    private final Map<TaskId, TaskContext> taskContexts = new ConcurrentHashMap();
+    private final Map<TaskId, TaskContext> taskContexts = new ConcurrentHashMap<>();
+
+    @GuardedBy("this")
+    private boolean isResourceOverCommit;
+    private boolean needsMemoryLimitsInitialized = true;
```

Comment:
Would this need to be volatile for `needsMemoryLimitsInitialized` to return an up to date value without synchronization with other threads?

Reply:
Ah, so you spotted it! I figured someone might. If it was more critical to avoid redundant initializations, you might indeed want to make it `volatile` (and I'm happy to do so if you feel strongly about it). The goal here is simply to avoid repeatedly contending for the `QueryContext` since all memory updates for the query also contend for that lock (very heavily, I might add). This is more of a "best effort" and it's not critical to definitely observe latest state since the failure mode is a redundant update (just like what happens today). Happy to change it that's what you'd prefer.

Reply:
OK, QueryContext is almost definitely safely published so we can rely on the default value.

Reply:
Decided to make it `volatile` anyway since clearly this was going to be a recurring feedback item.